KerbalStuff/blueprints/api.py

```diff
@@ -545,73 +560,65 @@ def create_list():
-    if not test_gameversion:
-        return { 'error': True, 'reason': 'Game version does not exist.' }, 400
-    game_version_id = test_gameversion.id
-    mod = Mod()
-    mod.user = current_user
-    mod.name = name
-    mod.game_id = game
-    mod.short_description = short_description
-    mod.description = default_description
-    mod.ckan = ckan
-    mod.license = license
+    if len(mod_name) > 100 \
+            or len(short_description) > 1000 \
+            or len(mod_licence) > 128:
+        return {'error': True, 'reason': 'Fields exceed maximum permissible length.'}, 400
+    game = None
+    if game_id:
+        game = Game.query.get(game_id)
+    elif game_short:
+        game = Game.query.filter(Game.short == game_short).first()
+    if not game:
+        return {'error': True, 'reason': 'Game does not exist.'}, 400
+    game_version_id = db.query(GameVersion.id) \
+        .filter(GameVersion.game_id == game.id) \
+        .filter(GameVersion.friendly_version == game_version) \
+        .first()
+    if not game_version_id:
+        return {'error': True, 'reason': 'Game version does not exist.'}, 400
     # Save zipball
-    filename = secure_filename(name) + '-' + secure_filename(version) + '.zip'
-    base_path = os.path.join(secure_filename(current_user.username) + '_' + str(current_user.id), secure_filename(name))
-    full_path = os.path.join(_cfg('storage'), base_path)
-    if not os.path.exists(full_path):
-        os.makedirs(full_path)
-    path = os.path.join(full_path, filename)
-    if os.path.isfile(path):
-        # We already have this version
-        # We'll remove it because the only reason it could be here on creation is an error
-        os.remove(path)
-    zipball.save(path)
-    if not zipfile.is_zipfile(path):
-        os.remove(path)
+    file_path = _save_mod_zipball(mod_name, friendly_version, zipball)
+    if not zipfile.is_zipfile(file_path):
         return {'error': True, 'reason': 'This is not a valid zip file.'}, 400
-    version = ModVersion(friendly_version=secure_filename(version),
+    version = ModVersion(friendly_version=friendly_version,
                          gameversion_id=game_version_id,
-                         download_path=os.path.join(base_path, filename))
+                         download_path=file_path)
+    # create the mod
+    mod = Mod(user=current_user,
+              name=mod_name,
+              short_description=short_description,
+              description=default_description,
+              license=mod_licence,
+              ckan=ckan in TRUE_STR,
+              game=game,
+              default_version=version)
     version.mod = mod
     # Save database entry
     db.add(mod)
     db.commit()
-    mod.default_version_id = version.id
-    db.commit()
-    set_game_info(Game.query.get(game))
-    if ckan:
+    set_game_info(game)
+    if mod.ckan:
```

Comment:
Maybe we should only allow CKAN for KSP mods?

Reply:
We should, but that's another issue that will be addressed separately.